### ISSUE TITLE
Magento Store Switch - 2 store views, 2 domains.  "The store that was requested wasn't found. Verify the store and try again."

### ISSUE THREAD

**Issue by kamzata** [magento/magento2#27540]
### Preconditions (*)

1. Magento 2.3.4 (fresh installation, self hosted)
2. 2 different domains
3. 1 website, 1 store, 2 store views.
4. Store view "en" associated at domain example.com from Magento backend
5. Store view "it" associated at domain example.it from Magento backend
6. Apache 2.4.38 (Debian 10)
7. _.htaccess_ configurated like that:
        `SetEnvIf Host ^(.*)\.example\.com MAGE_RUN_CODE=en`
        `SetEnvIf Host ^(.*)\.example\.com MAGE_RUN_TYPE=store`
        `SetEnvIf Host ^(.*)\.example\.it MAGE_RUN_CODE=it`
        `SetEnvIf Host ^(.*)\.example\.it MAGE_RUN_TYPE=store`

### Steps to reproduce (*)

1. Go to _example.com/my-beautiful-product.html_
2. Switch store view from _en_ to _it_

### Expected result (*)

1. Switch to _it_ store view at example.it/my-beautiful-product.html

### Actual result (*)

1. ![Schermata 2020-03-10 alle 19 25 08](https://user-images.githubusercontent.com/1103317/78193955-17a9eb80-747c-11ea-81de-a36c6b254d3c.png)

2. It takes me to _it_ store view home and it shows this error: **The store that was requested wasn't found. Verify the store and try again.**

3. I tried to hardcode the store view codes in */vendor/magento/module-store/Controller/Store/SwitchAction.php* at row 106 and the $requestedUrlToRedirect :

```
        ...
        public function execute()
        {
            $targetStoreCode = $this->_request->getParam(
                \Magento\Store\Model\StoreManagerInterface::PARAM_NAME
            );
            $fromStoreCode = $this->_request->getParam(
                '___from_store',
                $this->storeCookieManager->getStoreCodeFromCookie()
            );
    
            $requestedUrlToRedirect = 'https://example.it/my-beautiful-product.html';
            $redirectUrl = $requestedUrlToRedirect;
            // $requestedUrlToRedirect = $this->_redirect->getRedirectUrl();
            // $redirectUrl = $requestedUrlToRedirect;

            $error = null;
            try {
                $fromStore = $this->storeRepository->get('en');
                $targetStore = $this->storeRepository->getActiveStoreByCode('it');
                // $fromStore = $this->storeRepository->get($fromStoreCode);
                // $targetStore = $this->storeRepository->getActiveStoreByCode($targetStoreCode);
            } catch (StoreIsInactiveException $e) {
                $error = __('Requested store is inactive');
            } catch (NoSuchEntityException $e) {
                $error = __("The store that was requested wasn't found. Verify the store and try again.");
            }
            if ($error !== null) {
                $this->messageManager->addErrorMessage($error);
            } else {
                $redirectUrl = $this->storeSwitcher->switch($fromStore, $targetStore, $requestedUrlToRedirect);
            }

            $this->getResponse()->setRedirect($redirectUrl);
        }
        ...
```

Then I switched from Italian store view to the English one and **it worked!** So **it seems it does not able to get the correct values of $targetStoreCode, and $requestedUrlToRedirect**. Any ideas?

**Comment by kamzata** [comment]
@magento give me 2.4-develop instance - upcoming 2.4.x release

**Comment by magento-engcom-team** [comment]
Hi @kamzata. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @kamzata, here is your [Magento instance](https://i-27540-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-27540-2-4-develop.instances.magento-community.engineering/admin_dcbe](https://i-27540-2-4-develop.instances.magento-community.engineering/admin_dcbe)
Login: `fa6201db` Password: `d31e21ff8b69`
Instance will be terminated in up to 3 hours.

**Comment by kamzata** [comment]
I just made another installation on another similar environment and I got the same error. To see the error message you have to have Layout, Block HTML output, Page, and Translations cache disabled. Otherwise, you will bring on the other homepage store anyway (regardless where you were) but you will not able to see the error message "The store that was requested wasn't found. Verify the store and try again."

**Comment by kamzata** [comment]
I'm doing some debugging... it seems the problem is in _/vendor/magento/framework/HTTP/PhpEnvironment/Request.php_ at line 271.

```
    public function getAlias($name)
    {
        if (isset($this->aliases[$name])) {
            return $this->aliases[$name];
        }
        return null;
     }
```

It return `null` instead of returning the `___store` value. Same thing for `___from_store` value.

**Comment by kamzata** [comment]
I just tried on Magento 2.2.11 and it works great out of the box.

**Comment by kamzata** [comment]
I just tested on **Magento 2.3.0** and **it works partially out of the box**. Url rewrite has been created just for 1 store view. So, to make it works I had to add manually the url rewrites for the other store view. Then it worked.

**Comment by kamzata** [comment]
I just tested on **Magento 2.3.1** and **it shows the error**. The behaviour is just the same as for the 2.3.4 version.

**Comment by kamzata** [comment]
I'm trying to do some debugging comparing 2.2.11 and 2.3.4 versions... I noticed this:

 - Cookies `store` and `___from_store` in 2.2.11. are set while in 2.3.4 are not.

**Comment by kamzata** [comment]
I also noticed that in the 2.3.4 version while switching is just added to the url `/stores/store/switch/` whereas in 2.2.11 is added `/stores/store/switch/?___from_store=it&___store=en&uenc=aHR0cDovL2RldmxvYy50ZWNuaWNtYW4uY29tLz9fX19zdG9yZT1lbg%2C%2C`

**Comment by kamzata** [comment]
Great Scott! I think I find the bug. The problem is in the view... and exactly in `module-store/view/frontend/templates/switch/languages.phtml` at the line 28.

**WRONG**
```
<li class="view-<?= $block->escapeHtml($_lang->getCode()) ?> switcher-option">
    <a href="<?= $block->escapeUrl($block->getViewModel()->getTargetStoreRedirectUrl($_lang)) ?>">
        <?= $block->escapeHtml($_lang->getName()) ?>
    </a>
</li>
```

**CORRECT**
```
<li class="view-<?= $block->escapeHtml($_lang->getCode()) ?> switcher-option">
    <a href="#" data-post='<?= /* @noEscape */ $block->getTargetStorePostData($_lang) ?>'>
        <?= $block->escapeHtml($_lang->getName()) ?>
    </a>
</li>
```

...and now it works like a charm!

**Comment by kamzata** [comment]
I've just made a Pull request here: https://github.com/magento/magento2/pull/27880

**Comment by mrtuvn** [comment]
> I also noticed that in the 2.3.4 version while switching is just added to the url `/stores/store/switch/` whereas in 2.2.11 is added `/stores/store/switch/?___from_store=it&___store=en&uenc=aHR0cDovL2RldmxvYy50ZWNuaWNtYW4uY29tLz9fX19zdG9yZT1lbg%2C%2C`.
> 
> So, I tried to use it on Magento 2.3.4 pasted to the end of `http://example.com` like this `http://example.com/stores/store/switch/?___from_store=it&___store=en&uenc=aHR0cDovL2RldmxvYy50ZWNuaWNtYW4uY29tLz9fX19zdG9yZT1lbg%2C%2C` and it worked.

2.2.11 still use old implement url .  The params post to url use ? mark
In 2.3.0 magento use new implement url so no use ? anymore (for security i think)
My opinion both url still redirect to same one

**Comment by kamzata** [comment]
Anyway, it works partially out of the box with Magento 2.3.4. Url rewrite is created just for 1 store view. So, to make it works you have to add manually the url rewrites for the other store view.

**Comment by mrtuvn** [comment]
@kazamta seem core magento team already fixed issue in 2.4-develop branch code base internally
The root cause related Session lost after switching stores on different domains

**Comment by mrtuvn** [comment]
https://github.com/magento/magento2/commit/68770d182cfa085dc3fc2361d4bcd4fd06a30dc3
@kazamta here is commit

**Comment by kamzata** [comment]
> [68770d1](https://github.com/magento/magento2/commit/68770d182cfa085dc3fc2361d4bcd4fd06a30dc3)
> @kazamta here is commit

Just I tiny change :). Thanks @mrtuvn, good to know. I hope it will works well in the next release.

**Comment by pathakkrunal** [comment]
> Great Scott! I think I found the bug. The problem is in the view... and exactly in `module-store/view/frontend/templates/switch/languages.phtml` at the line 28.
> 
> **WRONG**
> 
> ```
> <li class="view-<?= $block->escapeHtml($_lang->getCode()) ?> switcher-option">
>     <a href="<?= $block->escapeUrl($block->getViewModel()->getTargetStoreRedirectUrl($_lang)) ?>">
>         <?= $block->escapeHtml($_lang->getName()) ?>
>     </a>
> </li>
> ```
> 
> **CORRECT**
> 
> ```
> <li class="view-<?= $block->escapeHtml($_lang->getCode()) ?> switcher-option">
>     <a href="#" data-post='<?= /* @noEscape */ $block->getTargetStorePostData($_lang) ?>'>
>         <?= $block->escapeHtml($_lang->getName()) ?>
>     </a>
> </li>
> ```
> 
> ...and now it works like a charm!

Work for me
Thank you.

**Comment by kamzata** [comment]
> > Great Scott! I think I found the bug. The problem is in the view... and exactly in `module-store/view/frontend/templates/switch/languages.phtml` at the line 28.
> > **WRONG**
> > ```
> > <li class="view-<?= $block->escapeHtml($_lang->getCode()) ?> switcher-option">
> >     <a href="<?= $block->escapeUrl($block->getViewModel()->getTargetStoreRedirectUrl($_lang)) ?>">
> >         <?= $block->escapeHtml($_lang->getName()) ?>
> >     </a>
> > </li>
> > ```
> > 
> > 
> > **CORRECT**
> > ```
> > <li class="view-<?= $block->escapeHtml($_lang->getCode()) ?> switcher-option">
> >     <a href="#" data-post='<?= /* @noEscape */ $block->getTargetStorePostData($_lang) ?>'>
> >         <?= $block->escapeHtml($_lang->getName()) ?>
> >     </a>
> > </li>
> > ```
> > 
> > 
> > ...and now it works like a charm!
> 
> Work for me
> Thank you.

Yeah, it works but probably in the "old way" as @mrtuvn said.

**Comment by ihor-sviziev** [comment]
Hi @kamzata,
Seems like this issue duplicates https://github.com/magento/magento2/issues/23755 and it was already resolved in 2.4-develop branch in https://github.com/magento/magento2/pull/26548.

Could you confirm that?

**Comment by kamzata** [comment]
> Hi @kamzata,
> Seems like this issue duplicates #23755 and it was already resolved in 2.4-develop branch in #26548 and included into 2.4.0 release.
> 
> Could you confirm that?

Now I'm using 2.4.0 version and it works fine.

**Comment by ihor-sviziev** [comment]
In this case I’m closing this issue
